File: extkernels/batchnorm.c
#include <math.h>
#include <stdint.h>
#include <stdlib.h>

void neuralops_batchnorm2d_fwd_mean(
    size_t batch_sz,
    size_t width,
    size_t height,
    size_t chan,
    const float *restrict in_buf,
    float *restrict mean)
{
  for (size_t a = 0; a < chan; a += 1) {
    mean[a] = 0.0f;
  }
  size_t p = 0;
  for (size_t idx = 0; idx < batch_sz; idx += 1) {
    for (size_t a = 0; a < chan; a += 1) {
      for (size_t y = 0; y < height; y += 1) {
        for (size_t x = 0; x < width; x += 1) {
          mean[a] += in_buf[p];
          p += 1;
        }
      }
    }
  }
  for (size_t a = 0; a < chan; a += 1) {
    mean[a] /= (float)(width * height * batch_sz);
  }
}

void neuralops_batchnorm2d_fwd_var(
    size_t batch_sz,
    size_t width,
    size_t height,
    size_t chan,
    const float *restrict in_buf,
    const float *restrict mean,
    float *restrict var)
{
  for (size_t a = 0; a < chan; a += 1) {
    var[a] = 0.0f;
  }
  size_t p = 0;
  for (size_t idx = 0; idx < batch_sz; idx += 1) {
    for (size_t a = 0; a < chan; a += 1) {
      float m = mean[a];
      for (size_t y = 0; y < height; y += 1) {
        for (size_t x = 0; x < width; x += 1) {
          float val = in_buf[p];
          var[a] += (val - m) * (val - m);
          p += 1;
        }
      }
    }
  }
  for (size_t a = 0; a < chan; a += 1) {
    var[a] /= (float)(width * height * batch_sz - 1);
  }
}

void neuralops_batchnorm2d_fwd_output(
    size_t batch_sz,
    size_t width,
    size_t height,
    size_t chan,
    const float *restrict in_buf,
    const float *restrict mean,
    const float *restrict run_mean,
    const float *restrict var,
    const float *restrict run_var,
    float *restrict out_buf,
    float gamma,
    float epsilon)
{
  size_t p = 0;
  for (size_t idx = 0; idx < batch_sz; idx += 1) {
    for (size_t a = 0; a < chan; a += 1) {
      /*float m = run_mean[a] * (1.0f - gamma) + mean[a] * gamma;
      float v = run_var[a] * (1.0f - gamma) + var[a] * gamma;*/
      float m = mean[a];
      float v = var[a];
      // FIXME(20160928): try rsqrtps intrinsic here.
      float rs = 1.0f / sqrtf(v + epsilon);
      for (size_t y = 0; y < height; y += 1) {
        for (size_t x = 0; x < width; x += 1) {
          out_buf[p] = (in_buf[p] - m) * rs;
          p += 1;
        }
      }
    }
  }
}

void neuralops_batchnorm2d_bwd_var(
    size_t batch_sz,
    size_t width,
    size_t height,
    size_t chan,
    const float *restrict in_buf,
    const float *restrict mean,
    const float *restrict run_mean,
    const float *restrict var,
    const float *restrict run_var,
    const float *restrict out_grad,
    float *restrict var_grad,
    float gamma,
    float epsilon)
{
  for (size_t a = 0; a < chan; a += 1) {
    var_grad[a] = 0.0f;
  }
  size_t p = 0;
  for (size_t idx = 0; idx < batch_sz; idx += 1) {
    for (size_t a = 0; a < chan; a += 1) {
      /*float m = run_mean[a] * (1.0f - gamma) + mean[a] * gamma;
      float v = run_var[a] * (1.0f - gamma) + var[a] * gamma;*/
      float m = mean[a];
      float v = var[a];
      float rs = -0.5f / ((v + epsilon) * sqrtf(v + epsilon));
      for (size_t y = 0; y < height; y += 1) {
        for (size_t x = 0; x < width; x += 1) {
          var_grad[a] += out_grad[p] * rs * (in_buf[p] - m);
          p += 1;
        }
      }
    }
  }
  /*for (size_t a = 0; a < chan; a += 1) {
    var_grad[a] *= -0.5f * gamma;
  }*/
}

void neuralops_batchnorm2d_bwd_mean(
    size_t batch_sz,
    size_t width,
    size_t height,
    size_t chan,
    const float *restrict in_buf,
    const float *restrict mean,
    const float *restrict run_mean,
    const float *restrict var,
    const float *restrict run_var,
    const float *restrict var_grad,
    const float *restrict out_grad,
    float *restrict mean_grad,
    float gamma,
    float epsilon)
{
  for (size_t a = 0; a < chan; a += 1) {
    mean_grad[a] = 0.0f;
  }
  size_t p = 0;
  for (size_t idx = 0; idx < batch_sz; idx += 1) {
    for (size_t a = 0; a < chan; a += 1) {
      /*float m = run_mean[a] * (1.0f - gamma) + mean[a] * gamma;
      float v = run_var[a] * (1.0f - gamma) + var[a] * gamma;*/
      float m = mean[a];
      float v = var[a];
      float rs = -1.0f / sqrtf(v + epsilon);
      float dv = var_grad[a];
      float c = -2.0f / ((float)(width * height * batch_sz - 1));
      for (size_t y = 0; y < height; y += 1) {
        for (size_t x = 0; x < width; x += 1) {
          mean_grad[a] += out_grad[p] * rs + dv * c * (in_buf[p] - m);
          p += 1;
        }
      }
    }
  }
  /*for (size_t a = 0; a < chan; a += 1) {
    mean_grad[a] *= -gamma;
  }*/
}

void neuralops_batchnorm2d_bwd_input(
    size_t batch_sz,
    size_t width,
    size_t height,
    size_t chan,
    const float *restrict in_buf,
    const float *restrict mean,
    const float *restrict run_mean,
    const float *restrict mean_grad,
    const float *restrict var,
    const float *restrict run_var,
    const float *restrict var_grad,
    const float *restrict out_grad,
    float *restrict in_grad,
    float gamma,
    float epsilon)
{
  size_t p = 0;
  for (size_t idx = 0; idx < batch_sz; idx += 1) {
    for (size_t a = 0; a < chan; a += 1) {
      /*float m = run_mean[a] * (1.0f - gamma) + mean[a] * gamma;
      float v = run_var[a] * (1.0f - gamma) + var[a] * gamma;*/
      float m = mean[a];
      float v = var[a];
      float dm = mean_grad[a];
      float cm = 1.0f / ((float)(width * height * batch_sz));
      float rs = 1.0f / sqrtf(v + epsilon);
      float dv = var_grad[a];
      float cv = 2.0f / ((float)(width * height * batch_sz - 1));
      for (size_t y = 0; y < height; y += 1) {
        for (size_t x = 0; x < width; x += 1) {
          in_grad[p] = out_grad[p] * rs + dm * cm + dv * cv * (in_buf[p] - m);
          p += 1;
        }
      }
    }
  }
}

File: src/kernels/batchnorm.rs
use kernels::ffi::*;

use densearray::prelude::*;

//use libc::{size_t};

pub struct BatchMean2dKernel {
  pub batch_sz: usize,
  pub dim:      (usize, usize, usize),
  pub mean:     Array1d<f32>,
  pub run_mean: Array1d<f32>,
}

pub struct BatchNorm2dKernel {
  pub batch_sz: usize,
  pub dim:      (usize, usize, usize),
  //pub gamma:    f32,
  pub epsilon:  f32,
  pub run_mean: Array1d<f32>,
  pub run_var:  Array1d<f32>,
  pub mean:     Array1d<f32>,
  pub mean_g:   Array1d<f32>,
  pub var:      Array1d<f32>,
  pub var_g:    Array1d<f32>,
}

impl BatchNorm2dKernel {
  pub fn new(batch_sz: usize, dim: (usize, usize, usize), epsilon: f32,) -> BatchNorm2dKernel {
    let run_mean = Array1d::zeros(dim.2);
    let mut run_var = Array1d::zeros(dim.2);
    run_var.as_view_mut().set_constant(1.0);
    let mean = Array1d::zeros(dim.2);
    let mean_g = Array1d::zeros(dim.2);
    let var = Array1d::zeros(dim.2);
    let var_g = Array1d::zeros(dim.2);
    BatchNorm2dKernel{
      batch_sz: batch_sz,
      dim:      dim,
      //gamma:    gamma,
      epsilon:  epsilon,
      run_mean: run_mean,
      run_var:  run_var,
      mean:     mean,
      mean_g:   mean_g,
      var:      var,
      var_g:    var_g,
    }
  }

  pub fn forward(&mut self, batch_sz: usize, in_buf: &[f32], out_buf: &mut [f32], gamma: f32) {
    assert!(batch_sz <= self.batch_sz);
    //unimplemented!();
    //self.mean.as_view_mut().set_constant(0.0);
    //self.var.as_view_mut().set_constant(0.0);
    assert_eq!(in_buf.len(), self.dim.flat_len() * batch_sz);
    assert_eq!(out_buf.len(), self.dim.flat_len() * batch_sz);
    unsafe { neuralops_batchnorm2d_fwd_mean(
        batch_sz,
        self.dim.0, self.dim.1, self.dim.2,
        in_buf.as_ptr(),
        self.mean.as_view_mut().as_mut_ptr(),
    ) };
    unsafe { neuralops_batchnorm2d_fwd_var(
        batch_sz,
        self.dim.0, self.dim.1, self.dim.2,
        in_buf.as_ptr(),
        self.mean.as_view().as_ptr(),
        self.var.as_view_mut().as_mut_ptr(),
    ) };
    unsafe { neuralops_batchnorm2d_fwd_output(
        batch_sz,
        self.dim.0, self.dim.1, self.dim.2,
        in_buf.as_ptr(),
        self.mean.as_view().as_ptr(),
        self.var.as_view().as_ptr(),
        out_buf.as_mut_ptr(),
        self.epsilon,
    ) };
  }

  pub fn backward(&mut self, batch_sz: usize, in_buf: &[f32], out_grad: &[f32], in_grad: &mut [f32], gamma: f32) {
    assert!(batch_sz <= self.batch_sz);
    //self.mean_grad.as_view_mut().set_constant(0.0);
    //self.var_grad.as_view_mut().set_constant(0.0);
    assert_eq!(in_buf.len(), self.dim.flat_len() * batch_sz);
    assert_eq!(out_grad.len(), self.dim.flat_len() * batch_sz);
    assert_eq!(in_grad.len(), self.dim.flat_len() * batch_sz);
    unsafe { neuralops_batchnorm2d_bwd_var(
        batch_sz,
        self.dim.0, self.dim.1, self.dim.2,
        in_buf.as_ptr(),
        self.mean.as_view().as_ptr(),
        self.var.as_view().as_ptr(),
        out_grad.as_ptr(),
        self.var_g.as_view_mut().as_mut_ptr(),
        self.epsilon,
    ) };
    unsafe { neuralops_batchnorm2d_bwd_mean(
        batch_sz,
        self.dim.0, self.dim.1, self.dim.2,
        in_buf.as_ptr(),
        self.mean.as_view().as_ptr(),
        self.var.as_view().as_ptr(),
        self.var_g.as_view().as_ptr(),
        out_grad.as_ptr(),
        self.mean_g.as_view_mut().as_mut_ptr(),
        self.epsilon,
    ) };
    unsafe { neuralops_batchnorm2d_bwd_input(
        batch_sz,
        self.dim.0, self.dim.1, self.dim.2,
        in_buf.as_ptr(),
        self.mean.as_view().as_ptr(),
        self.mean_g.as_view().as_ptr(),
        self.var.as_view().as_ptr(),
        self.var_g.as_view().as_ptr(),
        out_grad.as_ptr(),
        in_grad.as_mut_ptr(),
        self.epsilon,
    ) };
  }

  pub fn update(&mut self, gamma: f32) {
    if gamma == 1.0 {
      self.run_mean.as_mut_slice().copy_from_slice(&self.mean.as_slice());
      self.run_var.as_mut_slice().copy_from_slice(&self.var.as_slice());
    } else {
      self.run_mean.as_view_mut().vector_scale(1.0 - gamma);
      self.run_mean.as_view_mut().vector_add(gamma, self.mean.as_view());
      self.run_var.as_view_mut().vector_scale(1.0 - gamma);
      self.run_var.as_view_mut().vector_add(gamma, self.var.as_view());
    }
  }
}
